### PR TITLE
Refactors for issue #227 and pr https://github.com/metal/metal.js/pull/233

### PR DESCRIPTION
**Pull request by bryceosterhaus**
@mthadley, what do you think about this?

### CODE REVIEW

**Comment by mthadley** [comment]
🎉 Very nice! Just one more nitpick. We have docs for `oneOfType` and `shapeOf`, but is there any reason not to also document the rest of them? They are still methods on `Config`, after all.

File: packages/metal-state/src/Config.js [hunk 1]
```diff
@@ -27,6 +27,41 @@ import validators from './validators';
...
+	 */
+	oneOfType(validatorArray) {
+		validatorArray = validatorArray.map(
+			configObj => configObj.config.validator
+		);
+
+		return this.validator(validators.oneOfType(validatorArray));
+	},
+
+	/**
```
Comment (mthadley) [inline packages/metal-state/src/Config.js:111]: Missing a description here.

File: packages/metal-state/src/Config.js [hunk 2]
```diff
@@ -27,6 +27,41 @@ import validators from './validators';
...
+	arrayOf: setNestedValidators('arrayOf'),
+	bool: setPrimitiveValidators('bool'),
+	func: setPrimitiveValidators('func'),
+	instanceOf: setExplicitValueValidators('instanceOf'),
+	number: setPrimitiveValidators('number'),
+	object: setPrimitiveValidators('object'),
+	objectOf: setNestedValidators('objectOf'),
+	oneOf: setExplicitValueValidators('oneOf'),
+
+	/**
```
Comment (mthadley) [inline packages/metal-state/src/Config.js]: Missing a description here.

**Review by mthadley — CHANGES_REQUESTED** [review]

**Comment by bryceosterhaus** [comment]
Reason I didnt add docs for the rest of them is because the functions called are documented. I can add docs to everything, just wasnt sure.

**Comment by bryceosterhaus** [comment]
Updated docs

**Comment by mthadley** [comment]
@bryceosterhaus  I see. However, that doesn't mean that those methods are documented.

As an example, if you were to generate a documentation site from this file, those methods would not be documented.

**Comment by bryceosterhaus** [comment]
Added separate commit to further document methods in config.

**Comment by mthadley** [comment]
👍 👍 Really nice @bryceosterhaus 😍 😍 

Because of the usage for the more advanced validators is a little different, I'm hoping those extra docs will make it more obvious:

```js
MyComponent.STATE = {
  user: Config.shapeOf({
    /**
     * We actually have nested Config objects here, but only certain objects
     * make sense. In other words, you can't have `Config.value('Bryce')` here.
     * But `ConfigWithValidator` should be a nice hint.
     */
    name: Config.string(),
    age: Config.number()
  })
}
```

**Comment by yuchi** [comment]
Could you update the title of the issue? I had a mini heart attack thinking it would be a Metal refactor :)

**Comment by bryceosterhaus** [comment]
hahah yes I will update it, definitely don't want a full metal refactor

**Comment by jbalsas** [comment]
Hey @bryceosterhaus, @mthadley, is this ready to be merged? I can do so if you both are happy with it.

**Comment by bryceosterhaus** [comment]
@jbalsas lgtm

**Comment by mthadley** [comment]
@jbalsas I think it is good to go.

File: packages/metal-state/src/Config.js [hunk 3]
```diff
@@ -28,6 +28,104 @@ import validators from './validators';
  */
 const Config = {
 	/**
+	* A number, or a string containing a number.
```
Comment (jbalsas) [inline packages/metal-state/src/Config.js]: I guess this description isn't really what a `ConfigWithValidator` is?

**Review by jbalsas — CHANGES_REQUESTED** [review]

**Comment by bryceosterhaus** [comment]
@jbalsas, updated the description of `ConfigWithValidator`

**Pull request merged**